File: 백준/Gold/2170. 선 긋기/선 긋기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int result = 0;
        PriorityQueue<Node> pq = new PriorityQueue<>();
        StringTokenizer st;
        for(int n = 0; n < N; n++){
            st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());
            pq.add(new Node(start, end));
        }

        Node current = pq.poll();

        for(int n = 1; n < N; n++){
            Node next = pq.poll();
            // 겹치는 경우 end 확장
            if(current.isConnected(next)) current.end = Math.max(current.end, next.end);
            else{
                result += (current.end - current.start);
                current = next;
            }
        }
        result += (current.end - current.start);
        System.out.println(result);
    }

    static class Node implements Comparable<Node>{
        int start, end;

        public Node(int start, int end) {
            this.start = start;
            this.end = end;
        }

        public boolean isConnected(Node next){
            return this.end >= next.start;
        }

        @Override
        public int compareTo(Node o) {
            if(this.start == o.start) return this.end - o.end;
            return this.start - o.start;
        }
    }
}

File: 백준/Gold/2170. 선 긋기/선 긋기.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main(){
    val br = BufferedReader(InputStreamReader(System.`in`))
    var st: StringTokenizer? = null
    val N = br.readLine().toInt()
    val pq = PriorityQueue<Node>()
    var result: Int = 0

    for(n in 0 until N){
        st = StringTokenizer(br.readLine())
        pq.add(Node(st.nextToken().toInt(), st.nextToken().toInt()))
    }

    var current = pq.poll()

    while(!pq.isEmpty()){
        val next = pq.poll()
        if(current.isConnected(next)) current.end = maxOf(current.end, next.end)
        else {
            result += calc(current);
            current = next
        }
    }
    result += calc(current);
    println(result)
}

fun calc(node: Node): Int{
    return node.end - node.start
}

data class Node(var start: Int, var end: Int) : Comparable<Node>{
    fun isConnected(next: Node): Boolean{
        return this.end >= next.start
    }

    override fun compareTo(other: Node): Int {
        if(this.start == other.start) return this.end - other.end
        return this.start - other.start
    }
}